ProofCertificate wiring, Vaccination Repository Episode 3 (EXPOSUREAPP-6888, EXPOSUREAPP-6729)

d4rken opened the pull request:
Return of the proof certificate. :tm: 

@chris-cwa and did pair programming and wired up everything, breakthrough :partying_face: .

* Refactoring to re-use @chiljamgossow's decoding/parsing classes for the server part which shares a lot of the same data
* Additional unit tests for conversion/decoding/parsing
* Storage data decoding for proof certificates
* Polishing, polishing, polishing.


WIP:

- [x] More polishing
- [x] More tests

d4rken committed:
Vaccination data updates, draft 1: Skeletons.

d4rken committed:
Vaccination data updates, draft 2: Wiring calls to refresh and condition checks.

d4rken committed:
Vaccination data updates, draft 3: Unit tests.

d4rken committed:
Introduce unit test that checks application setup.

d4rken committed:
Improve logging.

d4rken committed:
Merge branch 'feature/vaccination-main' into feature/6733-vaccination-polling

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/VaccinatedPerson.kt

d4rken committed:
LINTs

d4rken committed:
Merge branch 'feature/vaccination-main' into feature/6733-vaccination-polling

chris-cwa committed:
wip2

chris-cwa committed:
fixed types

d4rken committed:
Wrap the RawCOSEObject to prevent confusion, and help with conversion.
typealias does not prevent missuse on compiler level.
Kotlin value classes are still experimental :(.

chris-cwa committed:
prod url

d4rken committed:
* Make the base45 a project utility class, align call pattern with base32 and base64. Additional unit tests.
* Align vaccination storage with COSE encoding/decoding. Unit tests.

d4rken committed:
Merge branch 'feature/vaccination-main' into feature/6733-vaccination-polling

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/VaccinatedPerson.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/VaccinationRepository.kt

d4rken committed:
wip

mtwalli committed:
Merge branch 'feature/vaccination-main' into feature/6733-vaccination-polling

d4rken committed:
wip

d4rken committed:
Change mockk init.

d4rken committed:
wip

d4rken committed:
Add missing task factory registration.

d4rken committed:
Merge branch 'feature/vaccination-main' into feature/6733-vaccination-polling

d4rken committed:
Merge branch 'feature/6733-vaccination-polling' into feature/6729-vaccination-repo-episode-2

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/VaccinatedPerson.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/qrcode/VaccinationQRCodeExtractor.kt

d4rken committed:
Resolve merge conflicts.

d4rken committed:
Merge branch 'feature/6729-vaccination-repo-episode-2' into feature/6888-proof-server

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/qrcode/VaccinationCertificateCOSEParser.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/qrcode/VaccinationCertificateV1Parser.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/qrcode/VaccinationQRCodeExtractor.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/storage/VaccinationContainer.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/VaccinationTestData.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/qrcode/VaccinationQRCodeExtractorTest.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/repository/VaccinationRepositoryTest.kt

d4rken committed:
Resolve merge conflicts.

d4rken committed:
Remove hardcoded proof check, works via UI now.

d4rken committed:
Merge branch 'feature/vaccination-main' into feature/6888-proof-server

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/qrcode/HealthCertificateCOSEDecoder.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/qrcode/VaccinationCertificateCOSEParser.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/qrcode/VaccinationCertificateQRCode.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/qrcode/VaccinationQRCodeExtractor.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/VaccinationRepository.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/storage/ProofContainer.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/storage/VaccinationContainer.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/storage/VaccinationStorage.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/server/proof/ProofCertificateCOSEParser.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/server/proof/ProofCertificateResponse.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/server/proof/VaccinationProofServer.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/RawCOSEObjectTest.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/VaccinationTestData.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/qrcode/VaccinationQRCodeExtractorTest.kt

d4rken committed:
Fix/Improve unit tests.

d4rken committed:
Additional unit tests and polishing, ZIP bomp handling.

d4rken committed:
LINTs

chiljamgossow commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/compression/ZLIBCompression.kt:
```diff
@@ -1,24 +1,34 @@
-            Timber.e(e)
-            throw InvalidInputException("Zlib decompression failed.")
+    fun decompress(input: ByteString, sizeLimit: Long = -1L): ByteString = try {
+        val inflaterSource = input.let {
```
Where are the zlib magic numbers?

d4rken replied:
A zlib(inflate) header check is part of `java.util.zip.Inflater`. I did not see a specific requirement for extra header checks in the tech spec, so I thought the default checks are sufficient.

Added a test for this:
https://github.com/corona-warn-app/cwa-app-android/pull/3144/commits/44a4ce5b7b4e95c7a980218ddfbd045e8031d79a#diff-b710f361f25e20d66b7b634de24f4e3017fd9e3f52fede9f4b3fa8a97fb22c21R18-R25

If the header doesn't match we get:

```java
de.rki.coronawarnapp.util.compression.InvalidInputException: ZLIB decompression failed.

	at de.rki.coronawarnapp.util.compression.ZLIBCompression.decompress(ZLIBCompression.kt:29)
	at de.rki.coronawarnapp.util.compression.ZLIBCompression.decompress$default(ZLIBCompression.kt:11)
	at de.rki.coronawarnapp.util.compression.ZLIBCompressionTest.basic decompression(ZLIBCompressionTest.kt:15)
(....)
Caused by: java.io.IOException: java.util.zip.DataFormatException: incorrect header check
	at okio.InflaterSource.readOrInflate(InflaterSource.kt:99)
	at de.rki.coronawarnapp.util.compression.ZLIBCompression.decompress(ZLIBCompression.kt:22)
	... 67 more
Caused by: java.util.zip.DataFormatException: incorrect header check
	at java.base/java.util.zip.Inflater.inflateBytesBytes(Native Method)
	at java.base/java.util.zip.Inflater.inflate(Inflater.java:385)
	at okio.InflaterSource.readOrInflate(InflaterSource.kt:79)
	... 68 more
```

chiljamgossow replied:
OK. It was in the reference implementation, that's why I included it.

d4rken committed:
Add test for magic header check.

d4rken committed:
Introduce ContainerPostProcessor.kt to allow injection of parsers into data containers.

d4rken committed:
Use DaggerVaccinationTestComponent to inject VaccinationQRCodeExtractor into VaccinationQRCodeExtractorTest.
Prevent DI object construction from accidentally differing from the one under test.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/storage/ContainerPostProcessor.kt:
```diff
@@ -0,0 +1,47 @@
+import javax.inject.Inject
+
+@Reusable
+class ContainerPostProcessor @Inject constructor(
+    private val vaccinationCertificateCOSEParser: VaccinationCertificateCOSEParser,
+    private val proofCertificateCOSEParser: ProofCertificateCOSEParser,
+) : TypeAdapterFactory {
```
Allows us to add extra data to classes right after deserialization. We don't have to hardcode the parsers anymore.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/certificate/RawCOSEObject.kt:
```diff
@@ -0,0 +1,82 @@
+import java.lang.reflect.Type
+import javax.inject.Inject
+
+data class RawCOSEObject(
```
Wrapper class around a `ByteString` representing raw COSE data.
`typealias` does not give us compiler checks, it's just a visual hint, extension functions are still on the underlying type.
Ideally we would have wanted `value class` here, but those are still experimental in Kotlin.

With the concrete type, we can have Retrofit/Json adapter for serialization.

mtwalli replied:
it is stable in 1.5.0, Luckily there is a PR for that https://github.com/corona-warn-app/cwa-app-android/pull/3130
I think you can use it here

d4rken replied:
That's awesome, I'll have a look after 1.5.0 is merged! Thanks!

d4rken commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/repository/VaccinationRepositoryTest.kt:
```diff
@@ -34,42 +24,42 @@ class VaccinationRepositoryTest : BaseTest() {
-            this.personIdentifier shouldBe VaccinationTestData.PERSON_A_VAC_1_CONTAINER.personIdentifier
-        }
-    }
+//    @BeforeEach
+//    fun setup() {
+//        MockKAnnotations.init(this)
+//
+//        every { timeStamper.nowUTC } returns nowUTC
+//
+//        every { valueSetsRepository.latestValueSet } returns flowOf(vaccinationValueSet)
+//
+//        coEvery { vaccinationProofServer.getProofCertificate(any()) } returns VaccinationTestData.PERSON_A_PROOF_1_RESPONSE
+//
+//        storage.apply {
+//            every { personContainers } answers { testStorage }
+//            every { personContainers = any() } answers { testStorage = arg(0) }
+//        }
+//    }
```
Follow up PR.

d4rken commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/VaccinationTestData.kt:
```diff
@@ -1,263 +1,121 @@
+import javax.inject.Inject
 
-object VaccinationTestData {
+class VaccinationTestData @Inject constructor(
+    private val vaccinationCertificateCOSEParser: VaccinationCertificateCOSEParser,
+    private val proofCertificateCOSEParser: ProofCertificateCOSEParser,
+) {
```
The helper class for `TestData` is injected because we want the test data including the parsers to actually evaluate it. Makes it a bit easier to use in other tests.

For each container we need to set the parser:

```kotlin
  val personAVac1Container = VaccinationContainer(
        scannedAt = Instant.ofEpochMilli(1620062834471),
        vaccinationCertificateCOSE = personAVac1COSE,
    ).apply {
        parser = vaccinationCertificateCOSEParser
    }
```

chiljamgossow reviewed (dismissed):
I checked mainly the refactoring of the QR code decoding. That has been done correctly.

mtwalli commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/storage/VaccinationContainer.kt:
```diff
@@ -1,48 +1,43 @@
 ) {
 
+    // Either set by [ContainerPostProcessor] or via [toVaccinationContainer]
+    @Transient lateinit var parser: VaccinationCertificateCOSEParser
```
this looks strange to me. Parser inside a data class ?! , can we improve this somehow?

d4rken replied:
I'm not 100% happy either, but so far the alternative seem  also not nice.
Would be another wrapper around this class and/or more logic related to this piece of data in a parent class.
If you have a good idea maybe we can have a call later.

mtwalli reviewed (dismissed):
Lgtm, One comment about the parser, I don't like to have it a data class. Probably there is another way to avoid that?

mtwalli committed:
Merge branch 'feature/vaccination-main' into feature/6888-proof-server

mtwalli committed:
Merge branch 'release/2.3.x' into feature/6888-proof-server

d4rken committed:
Merge branch 'release/2.3.x' into feature/6888-proof-server
